src/Core/Services/GraphQLRequestValidator.cs

```diff
@@ -0,0 +1,538 @@
+            // Case 2. Via another relationship where a referenced column by the parent of the current entity is a referencing column for a child of the current entity.
+
+            // Case 1: Remove from columnsToBeDerivedFromEntity, the columns which are autogenerated or
+            // have been provided a non-null value in the input for the current entity.
+            foreach (string columnToBeDerivedFromEntity in columnsToBeDerivedFromThisEntity)
+            {
+                if (sourceDefinition.Columns[columnToBeDerivedFromEntity].IsAutoGenerated)
+                {
+                    // The value for an autogenerated column is derivable.
+                    columnsToBeDerivedFromThisEntity.Remove(columnToBeDerivedFromEntity);
+                }
+                else if (backingColumnData.TryGetValue(columnToBeDerivedFromEntity, out IValueNode? value))
+                {
+                    if (value is null)
+                    {
+                        throw new DataApiBuilderException(
+                            message: $"Value cannot be null for referenced column.",
+                            statusCode: HttpStatusCode.BadRequest,
+                            subStatusCode: DataApiBuilderException.SubStatusCodes.BadRequest);
+                    }
+
+                    columnsToBeDerivedFromThisEntity.Remove(columnToBeDerivedFromEntity);
+                }
+            }
+
+            // Dictionary storing the mapping from relationship name to the set of referencing fields for all
+            // the relationships for which the current entity is the referenced entity.
+            Dictionary<string, HashSet<string>> columnsDerivedForRelationships = new();
+
+            // Dictionary storing the mapping from relationship name to the set of referenced fields for all
+            // the relationships for which the current entity is the referencing entity.
+            Dictionary<string, HashSet<string>> columnsToBeDerivedFromRelationships = new();
+
+            // Loop over all the relationship fields input provided for the current entity.
+            foreach (ObjectFieldNode fieldNode in objectFieldNodes)
+            {
+                (IValueNode? fieldValue, SyntaxKind fieldKind) = GraphQLUtils.GetFieldDetails(fieldNode.Value, context.Variables);
+                if (fieldKind is SyntaxKind.NullValue || GraphQLUtils.IsScalarField(fieldKind))
+                {
+                    // If the current field is a column/scalar field or has a null value, continue.
+                    continue;
+                }
+
+                string relationshipName = fieldNode.Name.Value;
+                string targetEntityName = runtimeConfig.Entities![entityName].Relationships![relationshipName].TargetEntity;
+                string? linkingObject = runtimeConfig.Entities![entityName].Relationships![relationshipName].LinkingObject;
+                if (!string.IsNullOrWhiteSpace(linkingObject))
```

Comment:
I think rather than using all those nested dictionary references altogether in the if check, it makes it easier to use a variable telling us what the if check is performed on.